robot_controllers/src/diff_drive_limiter.cpp

```diff
@@ -181,11 +194,87 @@ void DiffDriveLimiter::limit(double *limited_linear_velocity,
   // Limit right and left wheel velocites
   if (params_.scale_to_wheel_velocity_limits)
   {
-    // Scale wheel velocities together, to maintain curvature
-    double wheel_velocity = std::max(fabs(left), fabs(right));
-    double wheel_scale = scaleToLimit(wheel_velocity, params_.max_wheel_velocity);
-    left *= wheel_scale;
-    right *= wheel_scale;
+    // TODO(cleanup) Params to load; refactor out
+    static const double max_curvature_deviation_ = 0.05;
+    static const double min_velocity_ = 0.075;
+    static const double backoff_sigma_ = 0.01;
+    static const double min_backoff_ = 0.99;
+
+    // Compute last right/left wheel velocities from last desired commands
+    double last_left, last_right;
+    calcWheelVelocities(&last_left, &last_right, last_linear_velocity, last_angular_velocity);
+
+    // Compute desired curvature
+    const double desired_curvature = computeCuvature(desired_angular_velocity,
+                                                     desired_linear_velocity,
+                                                     1e-6);
+
+    // Are we past our motion threshold (ramped up)?
+    bool is_ramped = std::abs(feedback->right_wheel_velocity) > min_velocity_ &&
+                     std::abs(feedback->left_wheel_velocity)  > min_velocity_;
+
+    // Compute curvature from feedback
+    double current_curvature = desired_curvature;
+    if (feedback && is_ramped)
+    {
+      double current_linear_velocity  = (feedback->left_wheel_velocity +
+                                         feedback->right_wheel_velocity) / 2.0;
+      double current_angular_velocity = (feedback->right_wheel_velocity -
+                                         feedback->left_wheel_velocity) / params_.track_width;
+      current_curvature = computeCuvature(current_angular_velocity,
+                                          current_linear_velocity,
+                                          1e-2);
+
+      // Check curvature error after we've ramped up
+      double curvature_deviation = std::abs(current_curvature - desired_curvature);
+      if (curvature_deviation > max_curvature_deviation_)
+      {
+        // Compute backoff envelope based on curvature error
+        double dynamic_backoff = std::max(min_backoff_,
```

Comment:
I really think we need to avoid discontinuities when preforming limiting:

I think it is better to have the dynamic back-off start at 0.0 when curvature deviation is at max_curvature_deviation
```
     // Backoff starts at 1.0 when curvature_deviation = max_curvature_deviation 
     double curvature_deviation_delta = curvature_deviation_delta - max_curvature_deviation_; 
     double dynamic_backoff = std::exp(-backoff_sigma_ *  curvature_deviation_delta * curvature_deviation_delta));
```